numpy/core/tests/test_regression.py

```diff
@@ -2411,6 +2411,80 @@ def test_field_access_by_title(self):
             if HAS_REFCOUNT:
                 assert_(base <= sys.getrefcount(s))
 
+    @pytest.mark.slow
+    @pytest.mark.skipif(not sys.platform.startswith('linux'),
+                        reason="linux only")
+    @pytest.mark.skipif(sys.version_info < (3,4), reason="lacks venv")
+    @pytest.mark.skip("venv on travis is flakey")
```

Comment:
Skip the test entirely for now, maybe someday re-enable. It passes when run alone (`--k test_pickle_across_versions`) which gives me confidence that the actual fix works

Reply:
How about just look at the pickle header?

Reply:
Hmm, OK, looks like pickle itself determines the module, but it must be in the header somewhere. I wonder if the pickle can simply be searched for the module name? For instance
```
In [30]: a = array(1)

In [31]: pickle.dumps(a, protocol=4)
Out[31]: b'\x80\x04\x95\x96\x00\x00\x00\x00\x00\x00\x00\x8c\x1cnumpy.core._multiarray_umath\x94\x8c\x0c_reconstruct\x94\x93\x94\x8c\x05numpy\x94\x8c\x07ndarray\x94\x93\x94K\x00\x85\x94C\x01b\x94\x87\x94R\x94(K\x01)h\x03\x8c\x05dtype\x94\x93\x94\x8c\x02i8\x94K\x00K\x01\x87\x94R\x94(K\x03\x8c\x01<\x94NNNJ\xff\xff\xff\xffJ\xff\xff\xff\xffK\x00t\x94b\x89C\x08\x01\x00\x00\x00\x00\x00\x00\x00\x94t\x94b.'
```

Reply:
Looking at that, I wonder why the umath module is there? Maybe from the ndarray `__reduce__` method? AFAICT that method should not be called, being replaced by `__reduce_ex__`. I think we should look at both those methods.

Reply:
We should also test the different protocols.

Reply:
If you really want to test across versions, could also pickle some ndarray in an earlier version, get the hash of the result, and check that it agrees with the hash of the same data pickled with a latter version. The "correct" hashes can be stored in the test.

Reply:
For instance with current master and previous array
```
In [43]: hashlib.sha256(s).hexdigest()
Out[43]: 'cce278984d68606f0ceb4d23c298ebd8abe944adba1bc6b161f16c180e785fc3'
```

Reply:
The `array_reduce` function does look buggy, but it is unlikely it gets called unless some ndarray subclass does it explicitly.

Reply:
> I wonder why the umath module is there

The module is `_multiarray_umath`. 

I will write a test that checks the bytestring for the correct module name in all protocols. Tieing ourselves to the exact pickle bytestring or hash seems too restrictive going forward.